### ISSUE TITLE
Provider is slow when running from EC2 with other AWS providers

### ISSUE THREAD

**Issue by mohamed-haidara-cko** [phillbaker/terraform-provider-elasticsearch#281]
Hi, first of all, thank you for this provider. It is really helpful to set up an ES cluster :-) 

I have a similar issue to the one described on #124: when using the provider from an EC2 instance, the endpoint `/latest/api/token` is called multiple times to update a single resource making Terraform plan and apply really slow. I suspect with a throttling here but couldn't confirm it. **It took 40seconds to update a single resource.**

When running the same stack locally by using `aws_profile`, it works fine.

Here is my setup:
- An EC2 instance with a role that has access to the cluster
- Two AWS providers are used in the same stack:
  - The first one uses the EC2 assumed role
  - The second one assumes another role for cross-account access
-  An elasticsearch provider version 2.0.1:
   ```hcl
    provider "elasticsearch" {
       URL              = "https://custom.domain.com"
       aws_region       = "eu-west-1"
   }
   ```
   I disabled and enabled the health check but it's the same issue.
- Two ES resources in my stack: `elasticsearch_opensearch_role` and `elasticsearch_opensearch_roles_mapping` and other AWS resources
- ES version 7.10
 

<details>
  <summary>Here are the relevant logs when updating the elasticsearch_opensearch_roles_mapping :</summary>
  <pre>
2022-05-27T14:12:56.669Z [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/phillbaker/elasticsearch/2.0.1/linux_amd64/terraform-provider-elasticsearch_v2.0.1 args=[.terraform/providers/registry.terraform.io/phillbaker/elasticsearch/2.0.1/linux_amd64/terraform-provider-elasticsearch_v2.0.1]
2022-05-27T14:12:56.669Z [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/phillbaker/elasticsearch/2.0.1/linux_amd64/terraform-provider-elasticsearch_v2.0.1 pid=79
2022-05-27T14:12:56.669Z [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/phillbaker/elasticsearch/2.0.1/linux_amd64/terraform-provider-elasticsearch_v2.0.1
2022-05-27T14:12:56.679Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: configuring server automatic mTLS: timestamp=2022-05-27T14:12:56.679Z
2022-05-27T14:12:56.726Z [DEBUG] provider.terraform-provider-elasticsearch_v2.0.1: plugin address: address=/tmp/plugin625296976 network=unix timestamp=2022-05-27T14:12:56.726Z
2022-05-27T14:12:56.726Z [DEBUG] provider: using plugin: version=5
elasticsearch_opensearch_roles_mapping.engineers_role_mapping: Modifying... [id=engineers]
2022-05-27T14:12:56.808Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:12:56 [INFO] Using AWS: eu-west-1: timestamp=2022-05-27T14:12:56.808Z
2022-05-27T14:12:56.808Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:12:56 [[DEBUG]  DEBUG: Request ec2metadata/GetToken Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /latest/api/token HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Content-Length: 0
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 21600
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:12:56.808Z
elasticsearch_opensearch_roles_mapping.engineers_role_mapping: Still modifying... [id=engineers, 10s elapsed]
2022-05-27T14:13:06.842Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:06 [[DEBUG]  DEBUG: Request ec2metadata/GetToken Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /latest/api/token HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Content-Length: 0
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 21600
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:13:06.842Z
elasticsearch_opensearch_roles_mapping.engineers_role_mapping: Still modifying... [id=engineers, 20s elapsed]
2022-05-27T14:13:16.843Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [[DEBUG]  DEBUG: Request ec2metadata/GetMetadata Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /latest/meta-data/iam/security-credentials/ HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:13:16.843Z
2022-05-27T14:13:16.844Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [[DEBUG]  DEBUG: Response ec2metadata/GetMetadata Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 9
Accept-Ranges: none
Content-Type: text/plain
Date: Fri, 27 May 2022 14:13:16 GMT
Last-Modified: Fri, 27 May 2022 13:46:42 GMT
Server: EC2ws

-----------------------------------------------------]: timestamp=2022-05-27T14:13:16.844Z
2022-05-27T14:13:16.844Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [[DEBUG]  redactedRoleName]: timestamp=2022-05-27T14:13:16.844Z
2022-05-27T14:13:16.845Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [[DEBUG]  DEBUG: Request ec2metadata/GetMetadata Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /latest/meta-data/iam/security-credentials/redactedRoleName HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:13:16.844Z
2022-05-27T14:13:16.845Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [[DEBUG]  DEBUG: Response ec2metadata/GetMetadata Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 1442
Accept-Ranges: none
Content-Type: text/plain
Date: Fri, 27 May 2022 14:13:16 GMT
Last-Modified: Fri, 27 May 2022 13:46:42 GMT
Server: EC2ws

-----------------------------------------------------]: timestamp=2022-05-27T14:13:16.845Z
2022-05-27T14:13:16.845Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [[DEBUG]  {
  "Code" : "Success",
  "LastUpdated" : "2022-05-27T13:46:43Z",
  "Type" : "AWS-HMAC",
  "AccessKeyId" : "REDACTED",
  "SecretAccessKey" : "REDACTED",
  "Token" : "REDACTED",
  "Expiration" : "2022-05-27T19:49:36Z"
}]: timestamp=2022-05-27T14:13:16.845Z
2022-05-27T14:13:16.846Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [INFO] Pinging url to determine version https://custom.domain.com: timestamp=2022-05-27T14:13:16.846Z
2022-05-27T14:13:16.941Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:16 [INFO] ES version {REDACTED}: timestamp=2022-05-27T14:13:16.941Z
2022-05-27T14:13:17.016Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:17 [INFO] Using AWS: eu-west-1: timestamp=2022-05-27T14:13:17.016Z
2022-05-27T14:13:17.017Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:17 [[DEBUG]  DEBUG: Request ec2metadata/GetToken Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /latest/api/token HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Content-Length: 0
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 21600
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:13:17.017Z
elasticsearch_opensearch_roles_mapping.engineers_role_mapping: Still modifying... [id=engineers, 30s elapsed]
2022-05-27T14:13:27.077Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:27 [[DEBUG]  DEBUG: Request ec2metadata/GetToken Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /latest/api/token HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Content-Length: 0
/la-Ttl-Seconds: 21600
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:13:27.077Z
elasticsearch_opensearch_roles_mapping.engineers_role_mapping: Still modifying... [id=engineers, 40s elapsed]
2022-05-27T14:13:37.084Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:37 [[DEBUG]  DEBUG: Request ec2metadata/GetMetadata Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /latest/meta-data/iam/security-credentials/ HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:13:37.084Z
2022-05-27T14:13:37.085Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:37 [[DEBUG]  DEBUG: Response ec2metadata/GetMetadata Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 9
Accept-Ranges: none
Content-Type: text/plain
Date: Fri, 27 May 2022 14:13:37 GMT
Last-Modified: Fri, 27 May 2022 13:46:42 GMT
Server: EC2ws

-----------------------------------------------------]: timestamp=2022-05-27T14:13:37.085Z
2022-05-27T14:13:37.085Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:37 [[DEBUG]  redactedRoleName]: timestamp=2022-05-27T14:13:37.085Z
2022-05-27T14:13:37.085Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:37 [[DEBUG]  DEBUG: Request ec2metadata/GetMetadata Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /latest/meta-data/iam/security-credentials/redactedRoleName HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.43.21 (go1.16.15; linux; amd64)
Accept-Encoding: gzip

-----------------------------------------------------]: timestamp=2022-05-27T14:13:37.085Z
2022-05-27T14:13:37.086Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:37 [[DEBUG]  DEBUG: Response ec2metadata/GetMetadata Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 1442
Accept-Ranges: none
Content-Type: text/plain
Date: Fri, 27 May 2022 14:13:37 GMT
Last-Modified: Fri, 27 May 2022 13:46:42 GMT
Server: EC2ws

-----------------------------------------------------]: timestamp=2022-05-27T14:13:37.086Z
2022-05-27T14:13:37.087Z [INFO]  provider.terraform-provider-elasticsearch_v2.0.1: 2022/05/27 14:13:37 [[DEBUG]  {
  "Code" : "Success",
  "LastUpdated" : "2022-05-27T13:46:43Z",
  "Type" : "AWS-HMAC",
  "AccessKeyId" : "REDACTED",
  "SecretAccessKey" : "REDACTED",
  "Token" : "REDACTED",
  "Expiration" : "2022-05-27T19:49:36Z"
}]: timestamp=2022-05-27T14:13:37.086Z
elasticsearch_opensearch_roles_mapping.engineers_role_mapping: Modifications complete after 40s [id=engineers]
2022-05-27T14:13:37.108Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-05-27T14:13:37.110Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/phillbaker/elasticsearch/2.0.1/linux_amd64/terraform-provider-elasticsearch_v2.0.1 pid=79
2022-05-27T14:13:37.110Z [DEBUG] provider: plugin exited
  </pre>
</details>

Is there any way to speed up this?

**Comment by phillbaker** [comment]
Hello, thanks for opening an issue. Can you please include the following:
* terraform version
* provider version

> Two ES resources in my stack

How many resources are you managing in a single terraform run?

From the logs, it looks like `IMDSv2` is being used and a TTL of 21600 seconds is being used, so it doesn't look like throttling by STS.

It may be worth verifying network connection issues from the ec2 instance - can you try running the terraform plan from ec2 with the authentication method that worked well (`aws_profile`)?

**Comment by mohamed-haidara-cko** [comment]
Thanks for your response. 

- Terraform version: 1.0.9
- Provider version: 2.0.2
- 15 resources in the stack: 13 from AWS and 2 from this provider

I will keep you in touch regarding my tests from an EC2 instance.

**Comment by mohamed-haidara-cko** [comment]
Hi. I found the root cause of the issue. I was able to reproduce it and fix it. It's in fact a network issue with IMDSv2 since I was running Terraform inside a Docker container. A similar issue was raised in the Go SDK for AWS: https://github.com/aws/aws-sdk-go/issues/2972. A decrease in the timeout [here](https://github.com/phillbaker/terraform-provider-elasticsearch/blob/master/es/provider.go#L552) may mitigate the issue.

Anyway, the fix is to update the EC2 instance metadata `http_put_response_hop_limit` to something >= 2: `aws ec2 modify-instance-metadata-options --instance-id i-xxxxx --http-endpoint enabled --http-put-response-hop-limit 2`.

Feel free to close this issue if it's not relevant anymore and thanks for your help :-)
